lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -899,35 +899,132 @@ class MongoClientInterface {
-        const masterKey = formatMasterKey(objName, params.vFormat);
-        c.updateOne({
-            _id: masterKey,
-        }, {
+    putObjectNoVer(collection, bucketName, objName, value, params, log, cb) {
+        if (params?.needOplogUpdate) {
+            return this.putObjectNoVerWithOplogUpdate(collection, bucketName, objName, value, params, log, cb);
+        }
+        const key = formatMasterKey(objName, params.vFormat);
+        const putFilter = { _id: key };
+        return collection.updateOne(putFilter, {
             $set: {
-                _id: masterKey,
-                value: objVal,
+                _id: key,
+                value,
             },
         }, {
             upsert: true,
-        }).then(() => cb()).catch((err) => {
-            log.error('putObjectNoVer: error putting obect with no versioning', { error: err.message });
+        }).then(() => cb()).catch(err => {
+            log.error('internalPutObject: error putting object',
+                { bucket: bucketName, object: key, error: err.message });
             return cb(errors.InternalError);
         });
     }
 
+    /**
+     * Updates an object in a MongoDB collection without changing its version.
+     * If the object doesn't exist, it will be created (upsert is true for the second update operation).
+     * The operation is logged in the oplog.
+     *
+     * @param {Object} collection - The MongoDB collection to update the object in.
+     * @param {string} bucketName - The name of the bucket the object belongs to.
+     * @param {string} objName - The name of the object.
+     * @param {Object} value - The new value of the object.
+     * @param {Object} params - Additional parameters.
+     * @param {string} params.vFormat - object key format
+     * @param {string} params.originOp - origin operation
+     * @param {Object} log - The logger to use.
+     * @param {Function} cb - The callback function to call when the operation is complete.
+     * It is called with an error if there is an issue with the operation.
+     * @returns {void}
+     */
+    putObjectNoVerWithOplogUpdate(collection, bucketName, objName, value, params, log, cb) {
+        const key = formatMasterKey(objName, params.vFormat);
+        const putFilter = { _id: key };
+        // filter used when finding and updating object
+        const findFilter = Object.assign({
+            _id: key,
+            $or: [
+                { 'value.deleted': { $exists: false } },
+                { 'value.deleted': { $eq: false } },
+            ],
+        }, putFilter);
+        const updateDeleteFilter = Object.assign({
+            '_id': key,
+            'value.deleted': true,
+        }, putFilter);
+        return async.waterfall([
+            // Adding delete flag when getting the object
+            // to avoid having race conditions.
+            next => collection.findOneAndUpdate(findFilter, {
+                $set: updateDeleteFilter,
```

Comment:
overall, I wonder if the introduction of all these filters actually helps with readability: because it reduces locality and decouples the function call, filter and "update". What do you think?

Reply:
The fact that the filters are defined at the top may make it hard to follow indeed but it  reduces redundancy , and maintainability is easier than in the inline approach. We either make the code DRY (Don't Repeat Yourself) or easier to understand with the inline definition on each mongo op